src/accumulation.jl

```diff
@@ -38,30 +38,38 @@ end
     is_inplaceable_destination(x) -> Bool
 
 Returns true if `x` is suitable for for storing inplace accumulation of gradients.
-For arrays this boils down `x .= y` if will work to mutate `x`, if `y` is an appropriate
-tangent.
-Wrapper array types do not need to overload this if they overload `Base.parent`, and are
-`is_inplaceable_destination` if and only if their parent array is.
-Other types should overload this, as it defaults to `false`.
+For arrays this means `x .= y` will mutate `x`, if `y` is an appropriate tangent.
+
+Here "appropriate" means that both are real or both are complex,
+and that for structured matrices like `x isa Diagonal`, `y` shares this structure.
+
+Wrapper array types should overload this function if they can be written into.
+Before ChainRulesCore 1.16, it would guess `true` for most wrappers based on `parent`,
+but this is not safe, e.g. it will lead to an error with ReadOnlyArrays.jl. 
+
+There must always be a correct non-mutating path, so in uncertain cases,
+this function returns `false`.
 """
 is_inplaceable_destination(::Any) = false
-is_inplaceable_destination(::Array) = true
+
+is_inplaceable_destination(::DenseArray)= true
+is_inplaceable_destination(::DenseArray{<:Integer}) = false
+
 is_inplaceable_destination(::SparseVector) = true
 is_inplaceable_destination(::SparseMatrixCSC) = true
-is_inplaceable_destination(::BitArray) = true
-function is_inplaceable_destination(x::AbstractArray)
-    p = parent(x)
-    p === x && return false  # no parent
-    # basically all wrapper types delegate `setindex!` to their `parent` after some
-    # processing and so are mutable if their `parent` is.
-    return is_inplaceable_destination(p)
+
+function is_inplaceable_destination(x::SubArray)
+    alpha = is_inplaceable_destination(parent(x))
+    beta = x.indices isa Tuple{Vararg{ Union{Integer, Base.Slice, UnitRange}}}
+    return alpha && beta
```

Comment:
Is it clearer to do
```suggestion
    is_inplaceable_destination(parent(x)) || return falsse
    return  x.indices isa Tuple{Vararg{Union{Integer,Base.Slice,UnitRange}}}
```

I will trust your judgement

Reply:
I had a few variants. Maybe I don't like that this seems to imply you ought to check one before the other, but in fact it's just dealing with line length.